packages/block-library/src/navigation/index.php

```diff
@@ -643,11 +647,6 @@ function render_block_core_navigation( $attributes, $content, $block ) {
 		$inner_blocks = new WP_Block_List( $fallback_blocks, $attributes );
 	}
 
-	$parsed_blocks = parse_blocks( $navigation_post->post_content );
-	if ( block_core_navigation_block_contains_core_navigation( $parsed_blocks ) ) {
```

Comment:
Could the check be kept here, but changed to something like this?
```php
if ( block_core_navigation_block_contains_core_navigation( $inner_blocks ) ) {
```

I think that would cover more cases (i.e. where the inner blocks are serialized instead of pulled from the wp_navigation post).

`$inner_blocks` is a `WP_Block_List`, but it implements `Iterator`, so should still support loops. The values will be `WP_Block` objects, so it'd have to internally check for `$block->name` instead of `$block['blockName']`.

I think it'd be good to also add extra safety to `block_core_navigation_block_contains_core_navigation`, by having it return `false` when the `parsed_blocks` isn't set.

Reply:
> $inner_blocks is a WP_Block_List, but it implements Iterator, so should still support loops. The values will be WP_Block objects, so it'd have to internally check for $block->name instead of $block['blockName'].

Makes Sense! I have responded with dec0f03, but I would like to fix the unit test now.

> I think it'd be good to also add extra safety to block_core_navigation_block_contains_core_navigation, by having it return false when the parsed_blocks isn't set.

Do you mean to change it as follows?

```php
if ( ! isset( $parsed_blocks ) || block_core_navigation_block_contains_core_navigation( $inner_blocks ) ) {
	return '';
}
```

Reply:
Could this be solved simpler by moving this call up to be inside [the ref exists condition](https://github.com/WordPress/gutenberg/pull/47479/files#diff-90cb5b64c95087ded3c7bb160e38189c8499b0f14f61beb884c887452449fa12L604) ?

Reply:
> Could this be solved simpler by moving this call up to be inside [the ref exists condition](https://github.com/WordPress/gutenberg/pull/47479/files#diff-90cb5b64c95087ded3c7bb160e38189c8499b0f14f61beb884c887452449fa12L604) ?

I tried that approach with my [first commit](https://github.com/WordPress/gutenberg/pull/47479/commits/94b5bc3efb25b259eb153812450d8e51e3810b42). But, as a series of comments, it would be better not to limit it to only blocks with ref.

Reply:
> Do you mean to change it as follows?

@t-hamano Something like this is what I had in mind:

```php
function block_core_navigation_block_contains_core_navigation( $inner_blocks ) {
        if ( ! isset( $inner_blocks ) ) {
                return false;
        }

	foreach ( $inner_blocks as $block ) {
		if ( 'core/navigation' === $block->name ) {
			return true;
		}
		if ( block_core_navigation_block_contains_core_navigation( $block->inner_blocks ) ) {
			return true;
		}
	}
	return false;
}
```

Reply:
I have added an extra safety in 33780cf. This implementation is based on [the post template](https://github.com/WordPress/gutenberg/blob/b1d0a5524a9629220215dc6b525cad7f88465289/packages/block-library/src/post-template/index.php#L26).

I changed it to check when this function is called recursively in a foreach instead of at the beginning of the function. My understanding is that this problem can occur when the `block_core_navigation_block_contains_core_navigation()` is called recursively.